File: app/src/main/java/com/haivo/diana/View/MusicPresenter.java
package com.haivo.diana.View;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import androidx.annotation.Nullable;
import android.text.TextPaint;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import com.haivo.diana.R;

public class MusicPresenter extends View {
    public MusicPresenter(Context context) {
        super(context);
    }
    public MusicPresenter(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }
    public MusicPresenter(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }
    public MusicPresenter(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    Paint targetNotePaint = new TextPaint();
    Paint detectedNotePaint = new TextPaint();
    private String targetNote = "";
    private String detectedNote = "";
    private final static float SCALING_TOLERANCE = 0.05f;//To avoid changing size too much;
    private final static int UNDEFINED = -1;
    private float targetNoteScaleFactor = 1;
    private float detectedNoteScaleFactor = 1;
    private float targetNoteTextX, targetNoteTextY = UNDEFINED;
    private float detectedNoteTextX, detectedNoteTextY = UNDEFINED;

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        calculateTextSizeAndPosition();
        targetNotePaint.setColor(0xff00ddff);
        detectedNotePaint.setColor(getResources().getColor(R.color.colorRed));
        canvas.drawText(targetNote, targetNoteTextX, targetNoteTextY, targetNotePaint);
        canvas.drawText(detectedNote, targetNoteTextX, 800, detectedNotePaint);
    }

    /**
     * Divide canvas into 2 equal part vertically.
     * Inside each part, text should be in the middle and take half of the space
     */
    private void calculateTextSizeAndPosition() {
        // Target Note Calculation
        Rect bounds = new Rect();
        if (!TextUtils.isEmpty(targetNote)) {
            targetNotePaint.getTextBounds(targetNote, 0, targetNote.length(), bounds);
            float currentTextHeight = bounds.height();
            float currentTextWidth = bounds.width();
            float canvasWidth = getMeasuredWidth();
            targetNoteScaleFactor = (canvasWidth / 2) / currentTextWidth;
            float newTextWidth = 0;
            float newTextHeight = 0;
            if (Math.abs(targetNoteScaleFactor - 1) > SCALING_TOLERANCE
                || targetNoteTextX == UNDEFINED
                || targetNoteTextY == UNDEFINED) {
                targetNotePaint.setTextSize(targetNotePaint.getTextSize() * targetNoteScaleFactor);
                newTextWidth = currentTextWidth * targetNoteScaleFactor;
                newTextHeight = currentTextHeight * targetNoteScaleFactor;
                targetNoteTextX = (getMeasuredWidth() - newTextWidth) / 2;
                targetNoteTextY =
                    (float) getMeasuredHeight() / 2 - ((float) getMeasuredHeight() / 2 - newTextHeight / 2) / 2;
            }
            // Detected Note Calculation
            if (!TextUtils.isEmpty(detectedNote) && !detectedNote.equals("0")) {
                detectedNotePaint.getTextBounds(detectedNote, 0, detectedNote.length(), bounds);
                currentTextWidth = bounds.width();
                detectedNoteScaleFactor = (canvasWidth / 2) / currentTextWidth;
                if (Math.abs(detectedNoteScaleFactor - 1) > SCALING_TOLERANCE
                    || detectedNoteTextX == UNDEFINED
                    || detectedNoteTextY == UNDEFINED) {
                    detectedNotePaint.setTextSize(detectedNotePaint.getTextSize() * detectedNoteScaleFactor);
                    newTextWidth = currentTextWidth * detectedNoteScaleFactor;
                    detectedNoteTextX = (getMeasuredWidth() - newTextWidth) / 2;
                    detectedNoteTextY =
                        (float) getMeasuredHeight() - ((float) getMeasuredHeight() / 2 - newTextHeight / 2) / 2;
                }
            }
        }
    }

    public void setTargetNote(String note) {
        this.targetNote = note;
        invalidate();
    }

    public void setDetectedNote(String note) {
        this.detectedNote = note;
        invalidate();
    }
}

File: app/src/main/java/com/haivo/diana/View/MusicPresenter.kt
package com.haivo.diana.View

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.text.TextPaint
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import com.haivo.diana.R

class MusicPresenter : View {

    private var targetNotePaint: Paint = TextPaint()
    private var detectedNotePaint: Paint = TextPaint()
    private var targetNote = ""
    private var detectedNote = ""
    private var targetNoteScaleFactor = 1f
    private var detectedNoteScaleFactor = 1f
    private var targetNoteTextX: Float = 0.toFloat()
    private var targetNoteTextY = UNDEFINED.toFloat()
    private var detectedNoteTextX: Float = 0.toFloat()
    private var detectedNoteTextY = UNDEFINED.toFloat()

    constructor(context: Context) : super(context) {}
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {}

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        calculateTextSizeAndPosition()
        targetNotePaint.color = -0xff2201
        detectedNotePaint.color = resources.getColor(R.color.colorRed)
        canvas.drawText(targetNote, targetNoteTextX, targetNoteTextY, targetNotePaint)
        canvas.drawText(detectedNote, targetNoteTextX, 800f, detectedNotePaint)
    }

    /**
     * Divide canvas into 2 equal part vertically.
     * Inside each part, text should be in the middle and take half of the space
     */
    private fun calculateTextSizeAndPosition() {
        // Target Note Calculation
        val bounds = Rect()
        if (!TextUtils.isEmpty(targetNote)) {
            targetNotePaint.getTextBounds(targetNote, 0, targetNote.length, bounds)
            val currentTextHeight = bounds.height().toFloat()
            var currentTextWidth = bounds.width().toFloat()
            val canvasWidth = measuredWidth.toFloat()
            targetNoteScaleFactor = canvasWidth / 2 / currentTextWidth
            var newTextWidth = 0f
            var newTextHeight = 0f
            if (Math.abs(targetNoteScaleFactor - 1) > SCALING_TOLERANCE
                    || targetNoteTextX == UNDEFINED.toFloat()
                    || targetNoteTextY == UNDEFINED.toFloat()) {
                targetNotePaint.textSize = targetNotePaint.textSize * targetNoteScaleFactor
                newTextWidth = currentTextWidth * targetNoteScaleFactor
                newTextHeight = currentTextHeight * targetNoteScaleFactor
                targetNoteTextX = (measuredWidth - newTextWidth) / 2
                targetNoteTextY = measuredHeight.toFloat() / 2 - (measuredHeight.toFloat() / 2 - newTextHeight / 2) / 2
            }
            // Detected Note Calculation
            if (!TextUtils.isEmpty(detectedNote) && detectedNote != "0") {
                detectedNotePaint.getTextBounds(detectedNote, 0, detectedNote.length, bounds)
                currentTextWidth = bounds.width().toFloat()
                detectedNoteScaleFactor = canvasWidth / 2 / currentTextWidth
                if (Math.abs(detectedNoteScaleFactor - 1) > SCALING_TOLERANCE
                        || detectedNoteTextX == UNDEFINED.toFloat()
                        || detectedNoteTextY == UNDEFINED.toFloat()) {
                    detectedNotePaint.textSize = detectedNotePaint.textSize * detectedNoteScaleFactor
                    newTextWidth = currentTextWidth * detectedNoteScaleFactor
                    detectedNoteTextX = (measuredWidth - newTextWidth) / 2
                    detectedNoteTextY = measuredHeight.toFloat() - (measuredHeight.toFloat() / 2 - newTextHeight / 2) / 2
                }
            }
        }
    }

    fun setTargetNote(note: String) {
        this.targetNote = note
        invalidate()
    }

    fun setDetectedNote(note: String) {
        this.detectedNote = note
        invalidate()
    }

    companion object {
        private const val SCALING_TOLERANCE = 0.05f//To avoid changing size too much;
        private const val UNDEFINED = -1
    }
}
